### PR TITLE
API exposure - RESTful API and socket.io-based push Events

### LINKED ISSUE #10: Raiden RESTful API
*Expose the API of raiden to external clients.*

## General API Spec:

- **Setting up** the initial connection to the raiden network should be possible **manually** (by opening channels and depositing into them from single API methods)
- Additionally we want to provide a **(semi-)automatic setup** to the network from given configuration-parameters 
*(e.g. based on available assets, so that the node takes care of an optimal connectivity to the raiden network by opening and depositing into channels without any additional considerations from the user)*
- API calls and responses should include a specification of the **API-version** (up from v1.0), so that the API is **future-proof**. The app has to include all versions of the API in the future.

- **MvP**: this version of the API needs to provide all neccessary tools to make it possible to build a **Wallet style UI** (issue #285)

## Functionality
add following methods to `raiden-service.RaidenAPI`:

- `query_endpoint(raiden_address)` - for light clients
- `register_endpoint(endpoint, signature)`
- `get_channel_info()` - retrieve the timeout, deposit, etc

let the caller decide if the called method should be executed asynchronous or waiting
(== `gevent.Greenlet.wait()`)

- `asset_swap(..., wait=<bool>)`
- `expect_asset_swap(..., wait=<bool>)`
- `transfer(..., wait=<bool>)`

The asynchronous method should be the default option,
but when calling via REST the waiting method should be default. 
That's because in e.g. JavaScript one can work easily with Promises.

## Protocol:

- Endpoints of all methods in the `raiden_service.RaidenAPI` are accessible through http-requests and are structured complying to a **RESTful**-pattern. Resources are kept minimal:
     - (e.g. on `/api/channel`
     -`GET` - get channel info,
     -`POST` - open channel,
     -`DELETE` - close channel,
      -`UPDATE` - deposit to channel )
- All http-requests (*-> client-initiated*) are *'blocking'* and at least include information about the **success of the call** in their response message. That way we don't have to work with callback-ids on the client side (see #280 ) but rather can work with promises.
- For **non-client initiated** actions (e.g. incoming transfers, opened channels, channel closing etc.) there will be the **emission of events** from the node to the client
- **Filtering** of the events will be done **on the client side**, so all events will be put in a FIFO-queue and sequentially forwarded to the client
- We assume that **only 1 client per node** is connected
- Events and responses will be **serialized to JSON**, 'difficult' data encoding will be reused from the Ethereum protocol (pyethapp/pyethereum) where possible
    (e.g. hex encoding for addresses)

## Transport:

- Method-calls: **TCP** (http-requests)
- Events:   the client can choose between
    1) TCP (**polling** http-requests)
    2) TCP ( **long-polling**; *'fake' push notifications*)

- The **interface** for the API has to be **uniform**, independent of the chosen transport-layer

## Tasklist

- [x] [Expose blockchain events in the REST api](https://github.com/raiden-network/raiden/issues/467)
- [x] [Add http endpoint for exchange task](https://github.com/raiden-network/raiden/issues/465)
- [x] [Expose Raiden Rest API at Raiden start](https://github.com/raiden-network/raiden/issues/464) 
- [ ] [Expose Raiden internal events from the REST API](https://github.com/raiden-network/raiden/issues/466) --> #594  #466 
- [x] [Add rpc endpoints for `connect` and `leave` of token networks](https://github.com/raiden-network/raiden/issues/508)

### PR DESCRIPTION
**Pull request by ezdac**
this provides the initial ideas of the API-exposure.

### what this PR implements:
- `@expose` decorated methods can be called via http-requests (REST-style URLs)
- return-data from synchronous methods is included in the respond message `"result": data`
- asynchronous methods provide a callback-id in their respond message, but `"result": null`
- the callbacks for asynchronous methods are automatically generated in the `Dispatcher`-class
- the callback-id is later on used to publish an event to a websocket (`socket.io` protocol)

### what is still TO DO:
- all relevant state changes (*incoming transfers, channel closing, etc*) trigger such an event, which then gets passed to the connected client's websocket 

closes #10 
related #207

### CODE REVIEW

**Comment by hackaugusto** [comment]
Should we merge this PR or wait for the changes as defined in #10 ?

**Comment by ezdac** [comment]
let's wait for the changes in #10 , I will push to this PR directly and rename it accordingly

File: raiden/network/rpc/client.py [hunk 1]
```diff
@@ -3,8 +3,10 @@
 import rlp
 from ethereum import slogging
 from ethereum import _solidity
+from ethereum.exceptions import InvalidTransaction
 from ethereum.transactions import Transaction
 from ethereum.utils import denoms, int_to_big_endian, encode_hex, normalize_address
+<<<<<<< 4c31f51f14f0b23ec72f06c4605e9944fbe60a20
```
Comment (ezdac) [inline raiden/network/rpc/client.py]: note to self: don't get distracted while rebasing

File: raiden/raiden_service.py [hunk 2]
```diff
@@ -19,19 +20,26 @@
...
+    ChannelSecretRevealed,
+    ChannelSettled
+)
 
 log = slogging.get_logger(__name__)  # pylint: disable=invalid-name
 
 DEFAULT_REVEAL_TIMEOUT = 30
 DEFAULT_SETTLE_TIMEOUT = DEFAULT_REVEAL_TIMEOUT * 20
 
-EventListener = namedtuple('EventListener', ('event_name', 'filter_', 'translator'))
```
Comment (ezdac) [inline raiden/raiden_service.py]: **Careful**: this didn't make it in my pull-rebase from origin/master (somehow).
If I cannot rewind this easily, it has to be cherrypicked and reimplemented again!

**Comment by LefterisJP** [comment]
Superseded by both https://github.com/raiden-network/raiden/pull/388 and https://github.com/raiden-network/raiden/pull/408

**Pull request closed without merging**